component/discovery/discovery.go

```diff
@@ -7,15 +7,57 @@ import (
 	"time"
 
 	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/pkg/cluster"
 	"github.com/prometheus/prometheus/discovery"
 	"github.com/prometheus/prometheus/discovery/targetgroup"
 	"github.com/prometheus/prometheus/model/labels"
+	"github.com/rfratto/ckit/shard"
 )
 
 // Target refers to a singular discovered endpoint found by a discovery
 // component.
 type Target map[string]string
 
+// DistributedTargets uses the node's Lookup method to distribute discovery
+// targets when a Flow component runs in a cluster.
+type DistributedTargets struct {
```

Comment:
Idea only: I wonder if we can make this more generic....with generics? ;) Something like this pseudo code. So that we dont have to write specific ones for each type of object.

```go
// Declaration
Owns[T](key func(v T),values []T) []T

// Implementation
Owns[Target](func(t Target){
    return tgt.Labels().String()
},
targets)
```